enterprise_gateway/services/processproxies/yarn.py

```diff
@@ -94,6 +99,65 @@ def launch_process(self, kernel_cmd, **kwargs):
 
         return self
 
+    """submitting jobs to yarn queue and then checking till the jobs are in running state 
+           will lead to orphan jobs being created in some scenarios.
+           We take kernel_launch_timeout time and divide this into two parts.
+           if the queue is unavailable we take max 20% of the time to poll the queue periodically
+           and if the queue becomes available the rest of timeout is met in 80% of the remmaining 
+           time.  
+
+        """
+
+    def confirm_yarn_queue_avaibility(self, **kwargs):
+        """
+        confirms if the yarn queue has capacity to handle the resource requests that
+        will be sent to it.
+        if not, proper error messages are sent back for user experience
+        :param env_dict:
+        :return:
+        """
+
+        executer_memory = int(kwargs['env']['KERNEL_EXECUTOR_MEMORY'])
+        driver_memory = int(kwargs['env']['KERNEL_DRIVER_MEMORY'])
+
+        candidate_queue_name = str(kwargs['env']['KERNEL_QUEUE'])
+
+        node_label = str(kwargs['env']['KERNEL_NODE_LABEL'])
```

Comment:
We shouldn't require these values, nor do I think we can reliably come up with default values for these.  We should pick one of these that is required to enable this functionality.  I'm thinking `KERNEL_QUEUE`.  If that is provided, then the others must also be present for this logic to proceed.  If `KERNEL_QUEUE` is not provided, this logic is silently skipped.  However, if `KERNEL_QUEUE` is provided, but __any__ of the others are not, let's log a warning message indicating the missing others, but NOT raise an exception - just exit the method.

Probably better to use `**kwargs.get('env')` to avoid `KeyError` exceptions.

Reply:
Hi Kevin,

It's good we are discussing on this part of the code which needs to be generic and is most important.

To give you a background, we are internally using a capacity scheduler and the queue that we specify is actually split into two partitions internally. Using the value of node_label helps us identify which queue partition we should be looking at to query it's absolute used capacity.

Now if we are to generalize this, should we just look at the queue name and query it's used capacity on parent level ? Or do you suggest some other route ?

Reply:
Also from what i understand, in scenarios were we are using node labels to schedule the jobs, 
mentioning a particular node label only schedule jobs on nodes having that label, while if you don't mention any node label while submitting jobs, it will scheduled in the default partition.

In that scenario, assuming we have a queue name specified with/without node labels, we can look in the default/actual partition respectively inside that queue. 

But will this logic be generic enough ? I don't think people use a node partitioning to submit jobs in every scenario or do they ?

Reply:
Actually on further thoughts, there are various cases we will need to cover within capacity scheduling itself,

**Case 1 - Where user specifies queue name & node label**
This means the resources need to be requested exclusively on these node labels so we will have to query this node label for it's used capacity which is already implemented above.

**Case 2 - Where user specifies queue name without a node label**
According to yarn concepts, this will allow yarn to schedule job on nodes that are default as well as on nodes that are non-exclusive on which queue has access. This would require us to actually query all the partitions that are associated with this queue, then query this partition to understand if it is non-exclusive and then add the results up to determine a yes or no. (Work required)

**Case 3 - No node label concept is used.**
Have not dived into this, but this should be simple, we can just look into the used capacity of the queue itself.

**Case 4 -  It's not a capacity scheduler (some other scheduler )**
Open question

Reply:
Thanks for the information.  I am unfamiliar with YARN queueing.  Adding @lresende.

However, I think if we let the existence of `KERNEL_QUEUE_NAME` drive this, we might be okay (for these cases).  If `KERNEL_QUEUE_NAME` does not exist, I don't think we have any choice but to skip this code and proceed as we do today (which I assume is __Case 4__ for now).

How is __Case 3__ different from __Case 2__? In Case 3 you say _look into the used capacity of the queue itself_ implying that `KERNEL_QUEUE_NAME` was specified (unless there's some default queue that is implied by this).

If we require BOTH queue name and node label to enable this logic, then we address cases 1 and 4 (correct?).

I'm not sure how important Case 2 and 3 are.  Hoping others more familiar with YARN can chime in here.

Reply:
@kevin-bates I think this page can help us understand scheduling in YARN in a better way.

https://developer.ibm.com/hadoop/2017/03/10/yarn-node-labels/

Since we were using node labels, the PR actually tries to derive the remaining queue capacities given we have exclusive partitions(refer link). This covers Case 1. Case 2 is something in which the no nodel label is specified but the system still has node label partitioning.

Case 3 - which i have not researched as of now, is something where we have not set up yarn to have exlusive/non-exclusive partitions. It might be possible in this scenario, the solution is similar to Case 2 but we will have to dive a little.

Reply:
@IMAM9AIS - thank you for the link.

I don't think we should get caught up in how to cover all the cases at this point.  Let's identify either the `KERNEL_QUEUE_NAME` or `KERNEL_NODE_LABEL` (or both) as the "drivers" that enable this logic and get your primary case covered. If you can knock out others - great.  Other users that might have different use cases can adjust this method accordingly.  (Please add a link to the article in a nearby comment.)

We just need to ensure that not setting these values results in today's (default) behavior.

Reply:
@kevin-bates Agreed.

From what we understand, we will have to use both (queue name and node label as drivers ) for this logic. I will be fixing all comments and committing these changes soon (along with default behaviour logic) .

Reply:
I have added the changes. Please have a look @kevin-bates .